### ISSUE TITLE
Shared Node Idempotency and Pool Deletion Issues when using Consul SD

### ISSUE THREAD

**Issue by 0xHiteshPatel** [F5Networks/f5-appsvcs-extension#483]
### Environment
 * Application Services Version: 3.29.0-3
 * BIG-IP Version: 13.1.1.2.0.0.4

### Summary
When trying to delete pools using Consul service discovery one of the following happens:

1) Consistent - The pool removed from the declaration is not deleted on the device.  The trace shows a delete operation in the diff but the pool remains on the device
2) Intermittent - The operation fails with the error message "01070110:3: Node address '/Common/x.x.x.x' is referenced by a member of pool '/tenant/app/poolname'."

This issue is impacting a production environment and new deployments have been halted to prevent inconsistent configs and cascading errors in the customer environment.

### Steps To Reproduce
1. Submit the following declaration:
```json
{
    "class": "ADC",
    "schemaVersion": "3.29.0",
    "id": "9b1ab5b6-124d-4a66-9eb9-387eea3efaec",
    "updateMode": "selective",
    "deleteissue": {
        "class": "Tenant",
        "app": {
            "class": "Application",
            "template": "generic",
            "P_wiki": {
                "class": "Pool",
                "loadBalancingMode": "least-connections-member",
                "minimumMembersActive": 0,
                "members": [
                    {
                        "servicePort": 30034,
                        "addressDiscovery": "consul",
                        "autoPopulate": true,
                        "uri": "http://10.215.129.153:8082/catalog/nodes",
                        "updateInterval": 1,
                        "rejectUnauthorized": false,
                        "shareNodes": true
                    }
                ]
            },            
            "P_wiki2": {
                "class": "Pool",
                "loadBalancingMode": "least-connections-member",
                "minimumMembersActive": 0,
                "members": [
                    {
                        "servicePort": 30034,
                        "addressDiscovery": "consul",
                        "autoPopulate": true,
                        "uri": "http://10.215.129.153:8082/catalog/nodes",
                        "updateInterval": 1,
                        "rejectUnauthorized": false,
                        "shareNodes": true
                    }
                ]
            },
            "P_wiki3": {
                "class": "Pool",
                "loadBalancingMode": "least-connections-member",
                "minimumMembersActive": 0,
                "members": [
                    {
                        "servicePort": 30034,
                        "addressDiscovery": "consul",
                        "autoPopulate": true,
                        "uri": "http://10.215.129.153:8082/catalog/nodes",
                        "updateInterval": 1,
                        "rejectUnauthorized": false,
                        "shareNodes": true
                    }
                ]
            },
            "P_wiki4": {
                "class": "Pool",
                "loadBalancingMode": "least-connections-member",
                "minimumMembersActive": 0,
                "members": [
                    {
                        "servicePort": 30034,
                        "addressDiscovery": "consul",
                        "autoPopulate": true,
                        "uri": "http://10.215.129.153:8082/catalog/nodes",
                        "updateInterval": 1,
                        "rejectUnauthorized": false,
                        "shareNodes": true
                    }
                ]
            },
            "P_wiki5": {
                "class": "Pool",
                "loadBalancingMode": "least-connections-member",
                "minimumMembersActive": 0,
                "members": [
                    {
                        "servicePort": 30034,
                        "addressDiscovery": "consul",
                        "autoPopulate": true,
                        "uri": "http://10.215.129.153:8082/catalog/nodes",
                        "updateInterval": 1,
                        "rejectUnauthorized": false,
                        "shareNodes": true
                    }
                ]
            },
            "P_wiki6": {
                "class": "Pool",
                "loadBalancingMode": "least-connections-member",
                "minimumMembersActive": 0,
                "members": [
                    {
                        "servicePort": 30034,
                        "addressDiscovery": "consul",
                        "autoPopulate": true,
                        "uri": "http://10.215.129.153:8082/catalog/nodes",
                        "updateInterval": 1,
                        "rejectUnauthorized": false,
                        "shareNodes": true
                    }
                ]
            }            
        }
    }
}
```

2. Observe the following error response:
```json
    "results": [
        {
            "code": 200,
            "message": "success",
            "lineCount": 26,
            "host": "localhost",
            "tenant": "deleteissue",
            "runTime": 2154
        }
    ]
```

3. Delete a pool and POST:
```json
{
    "class": "ADC",
    "schemaVersion": "3.29.0",
    "id": "9b1ab5b6-124d-4a66-9eb9-387eea3efaec",
    "updateMode": "selective",
    "deleteissue": {
        "class": "Tenant",
        "app": {
            "class": "Application",
            "template": "generic",
            "P_wiki": {
                "class": "Pool",
                "loadBalancingMode": "least-connections-member",
                "minimumMembersActive": 0,
                "members": [
                    {
                        "servicePort": 30034,
                        "addressDiscovery": "consul",
                        "autoPopulate": true,
                        "uri": "http://10.215.129.153:8082/catalog/nodes",
                        "updateInterval": 1,
                        "rejectUnauthorized": false,
                        "shareNodes": true
                    }
                ]
            },            
            "P_wiki2": {
                "class": "Pool",
                "loadBalancingMode": "least-connections-member",
                "minimumMembersActive": 0,
                "members": [
                    {
                        "servicePort": 30034,
                        "addressDiscovery": "consul",
                        "autoPopulate": true,
                        "uri": "http://10.215.129.153:8082/catalog/nodes",
                        "updateInterval": 1,
                        "rejectUnauthorized": false,
                        "shareNodes": true
                    }
                ]
            },
            "P_wiki3": {
                "class": "Pool",
                "loadBalancingMode": "least-connections-member",
                "minimumMembersActive": 0,
                "members": [
                    {
                        "servicePort": 30034,
                        "addressDiscovery": "consul",
                        "autoPopulate": true,
                        "uri": "http://10.215.129.153:8082/catalog/nodes",
                        "updateInterval": 1,
                        "rejectUnauthorized": false,
                        "shareNodes": true
                    }
                ]
            },
            "P_wiki4": {
                "class": "Pool",
                "loadBalancingMode": "least-connections-member",
                "minimumMembersActive": 0,
                "members": [
                    {
                        "servicePort": 30034,
                        "addressDiscovery": "consul",
                        "autoPopulate": true,
                        "uri": "http://10.215.129.153:8082/catalog/nodes",
                        "updateInterval": 1,
                        "rejectUnauthorized": false,
                        "shareNodes": true
                    }
                ]
            },
            "P_wiki5": {
                "class": "Pool",
                "loadBalancingMode": "least-connections-member",
                "minimumMembersActive": 0,
                "members": [
                    {
                        "servicePort": 30034,
                        "addressDiscovery": "consul",
                        "autoPopulate": true,
                        "uri": "http://10.215.129.153:8082/catalog/nodes",
                        "updateInterval": 1,
                        "rejectUnauthorized": false,
                        "shareNodes": true
                    }
                ]
            }            
        }
    }
}
```

4.  Observe the result:
Results are attached due to their size.
[483-results.txt](https://github.com/F5Networks/f5-appsvcs-extension/files/6827505/483-results.txt)

5. The pool "P_wiki6" still exists on the device:
```
root@(asm1-np-vdca)(cfg-sync Disconnected)(Active)(/Common)(tmos)# list ltm pool /deleteissue/app/
Configuration Items:
  /deleteissue/app/P_wiki    /deleteissue/app/P_wiki2   /deleteissue/app/P_wiki3   /deleteissue/app/P_wiki4   /deleteissue/app/P_wiki5   /deleteissue/app/P_wiki6   
root@(asm1-np-vdca)(cfg-sync Disconnected)(Active)(/Common)(tmos)# list ltm pool /deleteissue/app/P_wiki6
ltm pool /deleteissue/app/P_wiki6 {
    load-balancing-mode least-connections-member
    members {
        10.241.197.117:30034 {
            address 10.241.197.117
        }
        10.241.197.120:30034 {
            address 10.241.197.120
        }
        10.241.197.187:30034 {
            address 10.241.197.187
        }
        10.241.197.189:30034 {
            address 10.241.197.189
        }
        10.241.197.38:30034 {
            address 10.241.197.38
        }
        10.241.197.39:30034 {
            address 10.241.197.39
        }
    }
    partition deleteissue
}
```

6. Sometimes, when performing the same operation this error is returned: "01070110:3: Node address '/Common/x.x.x.x' is referenced by a member of pool '/tenant/app/poolname'."

### Expected Behavior
The pool is properly deleted without error

### Actual Behavior
The pool is not deleted OR an error regarding the shared node is returned

**Comment by dstokesf5** [comment]
Thank you for filing this issue. I have added it to our internal product backlog as AUTOTOOL-2637.

**Comment by dstokesf5** [comment]
This issue has been resolved with the 3.30.0 release of AS3.
